### DISCUSSION TITLE
"Not enough memory resources are available to complete this operation"

### DISCUSSION THREAD

**Discussion by ChaseFlorom** [azuritecoin/OnnxDiffusersUI#1] [General]
Hey, I'm running a Radeon RX 5700 with 16GB and I'm getting a not enough  memory resources are available to complete this operation. Any ideas? I've had the CMD version of SD running and had it work.

Thanks!

**Comment by azuritecoin** [comment]
Are you running with a too many batch size or too large image? I find that on my RX 6900 XT 16GB i can run up to a batch size of 3 512x512 images or a single image size of 704x704. 768x512 is ok but slow. Try closing the UI prrogram and restarting it.

**Reply by ChaseFlorom** [reply]
It was totally a me problem for this one. I had messed with my comp the night before and one of my sticks of ram was partially inserted wrong.

**Comment by Dolyfin** [comment]
Having similar problem with vram. 12GB 5700XT, Batch size 1 and Iteration 1 seems to already be at the limit with 512x+ resolution. Also, it appears that I have to restart onnxUI.py every few runs because the VRAM would fill to max and never be cleared when idle. Eventually VRAM will be full enough to not allow to start unless exiting onnxUI.py which clears VRAM.

**Reply by azuritecoin** [reply]
Try a smaller image size of 384x384, or share more of your system RAM with the GPU. From my own experience, 16GB of total memory (VRAM + shared system memeory) should be enough for batch size of one 512x512 image.

**Comment by amida168** [comment]
I had this problem too. I have a laptop with hybrid GPUs.  I changed the Graphics Settings to solve this problem.

http://www.kombitz.com/2022/11/01/how-to-fix-stable-diffusion-not-enough-memory-resources-are-available-problem-on-windows-11/

**Comment by mikwee** [comment]
Same problem with my AMD Radeon RX 6600 and 8GB of RAM, in both 384x384 and 512x512. Batch size was 1 in both attempts.

**Reply by mikwee** [reply]
Update: 384x384 works with the CLI, 512x512 doesn't.

**Reply by azuritecoin** [reply]
Yeah, 8GB is the bare minimum that's needed to run stable diffusion. 512x512 needs more VRAM or shared system RAM.

**Reply by mikwee** [reply]
Yeah, honestly makes sense. 384x384 is impressive as is

**Comment by Tjololo12** [comment]
I run into this issue when trying to do two generations one after the other. It appears, looking at memory utilization, that the UI isn't releasing the memory from the previous batch. Is there a setting or something I need to enable to force a garbage collect after generation finishes? 6700XT

**Reply by azuritecoin** [reply]
The latest version of `onnxUI.py` has forced garbage collection after changing the pipeline. If you're willing to edit the script you could try adding `gc.collect()` at the begining and the end of `generate_click` function.

**Reply by WalterShomer** [reply]
which lines would that be? not programming savvy and tried a few combinations that didn't seem to work

**Reply by azuritecoin** [reply]
In `onnxUI.py` there's a function called `run_diffusers` around line 23. Scrolling down a bit, at line 42 you will see this `global pipe`.

Replace that line with the following. Make sure there's four spaces at the front of each line:
```python
    global pipe
    pipe = None
    gc.collect()
```

This will unload the pipeline and free up the memory. The downside is that everytime you run this it will have to reload the pipeline which takes some time.

**Comment by azuritecoin** [comment]
I've updated the UI to enable de-allocating the model and releasing the memory after each generation. The downside is that you need to reload the model at the start every time. Run the UI with the flag `python onnxUI.py --release-memory`

**Comment by IslandRhythms** [comment]
I've had a similar issue and I believe the problem is the addition of the scheduler. If you could add an option to the GUI where no scheduler is selected, I believe this may solve the problem.

**Reply by azuritecoin** [reply]
Do you mean use the same scheduler as the model instead of allowing the user to change it on the fly?

**Reply by IslandRhythms** [reply]
Yes. Its the only thing I can imagine is causing the problem if the GUI at its core is running the same code as the script.

**Reply by azuritecoin** [reply]
I've ran a test with and without the user specified scheduler and it doesn't affect the amount of memory used. I end up using 11.3/16 GPU memory after generating one 512x512 image for both scenarios.

**Reply by IslandRhythms** [reply]
I don't know what to tell you. I have a AMD RX 6800 XT and when I run the non GUI script it works fine. When I transition the GUI it tells me it doesn't have enough memory to run the operation. I've commented out the original line and put in a line where no scheduler is included as an arg and it works fine

**Reply by azuritecoin** [reply]
When did you setup stable diffusion on your PC? The current version of the UI doesn't need the scheduler flag when you call `onnxUI.py`

**Reply by IslandRhythms** [reply]
https://github.com/azuritecoin/OnnxDiffusersUI/blob/main/onnxUI.py#L217
What is the default scheduler that `from_pretrained` uses? I'll try again using that but when I changed that line to
 `pipe = OnnxStableDiffusionPipeline.from_pretrained(model_path, provider=provider)` I don't have an issue

**Reply by azuritecoin** [reply]
It depends on which model you load. eg. SD v1.5 will use PNDM by default. You can look in the <model name>\scheduler folder to see which scheduler your model uses.
